crypto/rand/rand_win.c

```diff
@@ -18,38 +18,102 @@
 # endif
 
 # include <windows.h>
-/* On Windows Vista or higher use BCrypt instead of the legacy CryptoAPI */
-# if defined(_MSC_VER) && _MSC_VER > 1500 /* 1500 = Visual Studio 2008 */ \
-     && defined(_WIN32_WINNT) && _WIN32_WINNT >= 0x0600
-#  define USE_BCRYPTGENRANDOM
+# include <wincrypt.h>
+# include "internal/thread_once.h"
+
+# ifndef STATUS_SUCCESS
+#  define STATUS_SUCCESS ((NTSTATUS)0x00000000L)
+# endif
+
+# ifndef GET_MODULE_HANDLE_EX_FLAG_PIN
+# define GET_MODULE_HANDLE_EX_FLAG_PIN 0x00000001
+# endif
+
+# ifndef BCRYPT_USE_SYSTEM_PREFERRED_RNG
+# define BCRYPT_USE_SYSTEM_PREFERRED_RNG 0x00000002
 # endif
 
-# ifdef USE_BCRYPTGENRANDOM
-#  include <bcrypt.h>
-#  pragma comment(lib, "bcrypt.lib")
-#  ifndef STATUS_SUCCESS
-#   define STATUS_SUCCESS ((NTSTATUS)0x00000000L)
-#  endif
-# else
-#  include <wincrypt.h>
+/* prototype of GetModuleHandleExA used to pin bcrypt.dll in memory */
+typedef BOOL (WINAPI *GetModuleHandleExAFn)(
+    DWORD        dwFlags,
+    LPCSTR     lpModuleName,
+    HMODULE*    phModule
+);
+
+/* prototype of BCryptGenRandom function from bcrypt.dll */
+typedef long (WINAPI *BCryptGenRandomFn)(
+    void* hAlgorithm,
+    unsigned char* pbBuffer,
+    unsigned long cbBuffer,
+    unsigned long dwFlags
+);
+
+static BCryptGenRandomFn BCryptGenRandomPtr = NULL;
+static CRYPTO_ONCE load_bcrypt_dll = CRYPTO_ONCE_STATIC_INIT;
+DEFINE_RUN_ONCE_STATIC(do_load_bcrypt_dll)
+{
+    HMODULE hBCryptDll = NULL;
+    char system32_path[MAX_PATH];
+    GetModuleHandleExAFn GetModuleHandleExAPtr = NULL;
+    
+    /* We get pointer to GetModuleHandleExA dynamically in case it's not 
+     * defined in the compiler's headers
+     */
+    GetModuleHandleExAPtr = 
+        (GetModuleHandleExAFn) GetProcAddress (
+            GetModuleHandleA ("kernel32.dll"), 
+            "GetModuleHandleExA");
+    if (GetModuleHandleExAPtr != NULL) {
+        /* we load bcrypt.dll using absolute path to protect from
+         * possible dll hijacking since bcrypt.dll is not part of 
+         * the KnownDlls list protected by Windows from such attacks
+         */
+        if (GetSystemDirectoryA (system32_path, MAX_PATH) 
+                && (strlen(system32_path) < (MAX_PATH - 11))) {
+            /* we have enough room for "\bcrypt.dll" */
+            strcat (system32_path, "\\bcrypt.dll");
+        }
+        else {
+            strcpy (system32_path, "C:\\Windows\\System32\\bcrypt.dll");
```

Comment:
Why we need this?

Reply:
This is to handle the case of `GetSystemDirectoryA` call failing. This should never happen of course. As an alternative we can also consider that the system is broken in such event and simply not use bcrypt.dll instead of using the hard-coded path.